File: 2037/b.py
from collections import Counter

def solve(length, mixed_grids):
    k = length - 2
    counts = Counter(mixed_grids)
    mixed_grids_set = set(mixed_grids)

    for a in range(1, int(k ** 0.5) + 1):
        if k % a == 0:
            b = k // a
            if a in mixed_grids_set and b in mixed_grids_set:
                if a == b and counts[a] < 2:
                    continue
                return f"{a} {b}"


count = int(input())

for i in range(count):
    length = int(input())
    mixed_grids = list(map(int, input().split()))
    print(solve(length, mixed_grids))

File: 2037/b.cpp
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <cmath>

using namespace std;

pair<int, int> solve(int length, const vector<int>& mixed_grids) {
    int k = length - 2;
    unordered_map<int, int> counts;
    unordered_set<int> mixed_grids_set;

    for (int num : mixed_grids) {
        counts[num]++;
        mixed_grids_set.insert(num);
    }

    for (int a = 1; a <= sqrt(k); ++a) {
        if (k % a == 0) {
            int b = k / a;
            if (mixed_grids_set.count(a) && mixed_grids_set.count(b)) {
                if (a == b && counts[a] < 2) {
                    continue;
                }
                return {a, b};
            }
        }
    }
    return {-1, -1};
}

int main() {
    int count;
    cin >> count;

    for (int i = 0; i < count; ++i) {
        int length;
        cin >> length;
        vector<int> mixed_grids(length);
        for (int j = 0; j < length; ++j) {
            cin >> mixed_grids[j];
        }
        pair<int, int> result = solve(length, mixed_grids);
        if (result.first != -1) {
            cout << result.first << " " << result.second << endl;
        }
    }

    return 0;
}
